rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java

```diff
@@ -97,24 +94,27 @@ public List<ToolItemRubricAssociation> findByItemIdPrefix(String toolId, String
         return session.createQuery(query).list();
     }
 
-    public ToolItemRubricAssociation save(ToolItemRubricAssociation assoc) {
-
-        Rubric rubric = assoc.getRubric();
-        List<ToolItemRubricAssociation> associations = rubric.getAssociations();
-        int index = associations.indexOf(assoc);
-
-        if (index != -1) {
-            associations.set(index, assoc);
-        } else {
-            associations.add(assoc);
-        }
+    public ToolItemRubricAssociation save(ToolItemRubricAssociation association) {
+        ToolItemRubricAssociation mergedAssociation = null;
 
-        rubric = rubricRepository.save(rubric);
-        if (index != -1) {
-            return rubric.getAssociations().get(index);
+        Long associationId = association.getId();
+        if (associationId != null) {
+            // updating an existing association we can merge it and return it as nothing changes on the rubric
+            mergedAssociation = (ToolItemRubricAssociation) sessionFactory.getCurrentSession().merge(association);
         } else {
-            return rubric.getAssociations().get(rubric.getAssociations().size() - 1);
+            // adding a new association
+            if (association.getRubric() != null) {
+                Long rubricId = association.getRubric().getId();
+                Rubric rubric = rubricRepository.getById(rubricId);
+                if (rubric != null) {
+                    rubric.getAssociations().add(association);
+                    rubricRepository.save(rubric);
+                    // query for the new association
+                    mergedAssociation = findByItemIdAndRubricId(association.getItemId(), rubricId).orElse(null);
```

Comment:
This line implies that a null can be returned. I know it probably never will be, but the code reads like that.

Reply:
Correct a null is definitely possible to be returned, there are checks to ensure the association has a rubric. Should an association be persisted without a rubric? maybe the correct thing is to raise an exception here vs returning null?